### PR TITLE
feat: waku store sync 2.0 storage and tests

### PR DESCRIPTION
**Pull request by SionoiS**
# Description

Second PR for new Waku store sync 2.0

Include everything Storage related and tests

[Specification](https://github.com/waku-org/specs/blob/feat--waku-sync-2/standards/core/sync.md)
[Research issue](https://github.com/waku-org/research/issues/102)

# Changes

- [x] Storage interface
- [x] Storage impl.
- [x] tests

N.B. Cannot be merged before #3213
Followed by #3216 

# Issue
- https://github.com/waku-org/nwaku/issues/3221

### CODE REVIEW

**Review by jm-clius — APPROVED** [review]
Right, went through this in some detail. lgtm! Good job :). Would it be useful to have a set of unit tests for each public method (`insert`, `batchInsert`, `prune`) to test the boundary conditions at each?

**Comment by SionoiS** [comment]
> Right, went through this in some detail. lgtm! Good job :). Would it be useful to have a set of unit tests for each public method (`insert`, `batchInsert`, `prune`) to test the boundary conditions at each?

Yes, I will add some.

File: waku/waku_store_sync/storage/storage.nim [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import std/math, results, chronos
+
+import ../../waku_core/time, ../common
+
+type SyncStorage* = ref object of RootObj
+
+method insert*(
+    self: SyncStorage, element: ID
+): Result[void, string] {.base, gcsafe, raises: [].} =
+  discard
```
Comment (SionoiS) [inline waku/waku_store_sync/storage/storage.nim:10]: Maybe `insert` should be idempotent?

File: waku/waku_store_sync/storage/seq_storage.nim [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+  if upper < 1:
+    # entire range is before any of our elements
+    return none(Slice[int])
+
+  if lower >= self.elements.len:
+    # entire range is after any of our elements
+    return none(Slice[int])
+
+  return some(lower ..< upper)
```
Comment (Ivansete-status) [inline waku/waku_store_sync/storage/seq_storage.nim:89]: Should the returned `Slice` not include the upper too?
```suggestion
  return some(lower .. upper)
```

File: waku/waku_store_sync/storage/seq_storage.nim [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+  if ourFingerprint == inputFingerprint:
+    output.ranges.add((inputBounds, skipRange))
+    return
+
+  let slice =
+    if idxSlice.isNone():
+      output.ranges.add((inputBounds, itemSetRange))
+      let state = ItemSet(elements: @[], reconciled: true)
+      output.itemSets.add(state)
```
Comment (Ivansete-status) [inline waku/waku_store_sync/storage/seq_storage.nim:116]: I'm a bit puzzled about these two lines :)
Are we trying to "return" an empty `SyncPayload`?

File: waku/waku_store_sync/storage/seq_storage.nim [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    else:
+      idxSlice.get()
+
+  if slice.len <= self.lengthThreshold:
+    output.ranges.add((inputBounds, itemSetRange))
+    let state = ItemSet(elements: self.elements[slice], reconciled: false)
+    output.itemSets.add(state)
+    return
+
+  for partitionBounds in equalPartitioning(inputBounds, self.partitionCount):
```
Comment (Ivansete-status) [inline waku/waku_store_sync/storage/seq_storage.nim:127]: I think it is better to invoke `equalPartitioning` once and later iterate over the result. I believe we should only use that approach if `equalPartitioning` defined some kind of `iterator` with `yield` inside.

File: waku/waku_store_sync/storage/seq_storage.nim [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      if not inputItemSet.reconciled:
+        output.ranges.add((inputBounds, itemSetRange))
+        let state = ItemSet(elements: @[], reconciled: true)
+        output.itemSets.add(state)
+      else:
+        output.ranges.add((inputBounds, skipRange))
+
+      return
+    else:
+      idxSlice.get()
```
Comment (Ivansete-status) [inline waku/waku_store_sync/storage/seq_storage.nim:173]: I see some room for encapsulation as apparently we have similar approach two or three times

**Review by Ivansete-status — APPROVED** [review]
Thanks for it :partying_face: 
Approving to not block
I just added some nitpick comments ;P

File: waku/waku_store_sync/storage/storage.nim (hunk 1 above)
Comment (SionoiS) [reply waku/waku_store_sync/storage/storage.nim:10]: Currently if you can insert with the same element the first call would be `ok()` but if you call again `err()` since the element was already present.

An idempotent function would always insert the same element at the same index and would result in no side effect when calling the same func with the same params multiple times.

File: waku/waku_store_sync/storage/seq_storage.nim (hunk 4 above)
Comment (SionoiS) [reply waku/waku_store_sync/storage/seq_storage.nim:127]: > I think it is better to invoke `equalPartitioning` once and later iterate over the result.

Is that not how it works under the hood?

File: waku/waku_store_sync/storage/seq_storage.nim (hunk 2 above)
Comment (SionoiS) [reply waku/waku_store_sync/storage/seq_storage.nim:89]: :thinking: I'm pretty sure that the `Slice[ID]` exclude the upper bound so we do the same.

File: waku/waku_store_sync/storage/seq_storage.nim [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    self: SeqStorage, bounds: Slice[ID]
+): Fingerprint {.raises: [].} =
+  let idxSliceOpt = self.findIdxBounds(bounds)
+  return self.fingerprinting(idxSliceOpt)
+
+proc processFingerprintRange*(
+    self: SeqStorage,
+    inputBounds: Slice[ID],
+    inputFingerprint: Fingerprint,
+    output: var SyncPayload,
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_store_sync/storage/seq_storage.nim:101]: Well there's no way for the proc to fail.

Also, the `var payload` could have been local to the proc but the caller has a loop so it's inconvenient.

File: waku/waku_store_sync/storage/seq_storage.nim (hunk 3 above)
Comment (SionoiS) [reply waku/waku_store_sync/storage/seq_storage.nim:116]: Yes it's a valid case where one side have an empty range.

Peer 1:
range A to F
fingerprint 0x53...

Peer 2:
range A to F
no elements :(
send back an empty item set

File: waku/waku_store_sync/storage/seq_storage.nim (hunk 5 above)
Comment (SionoiS) [reply waku/waku_store_sync/storage/seq_storage.nim:173]: They are slightly different. One has a `return` the other `continue`
